sys/mips/mips/trap.c

```diff
@@ -1156,6 +1191,15 @@ trap(struct trapframe *trapframe)
 		if (KTRPOINT(td, KTR_CEXCEPTION))
 			ktrcexception(trapframe);
 #endif
+
+		if (try_emulate_capdirty(trapframe, NULL,
+					 &p->p_vmspace->vm_pmap) == 0) {
+			if (!usermode) {
```

Comment:
We know we're in a T_USER case, and there's no fallthrough (nor any gotos) to get us here otherwise, so this should always be true. You're also not passing `kernel_pmap` anyway for `!usermode`, and maybe this means you can get rid of the `if (kpmap)` check? Though maybe you can still trigger CP2 TLB faults for kernel VAs from usermode...

Reply:
Good catch.

I'm not sure whether we can get a C2E from user PCs on kernel VAs (I'd hope the segment checks fire first), but it seems like a reasonable defense in depth to only pass the kpmap in when coming from a kernel PC (in the prior call to try_emulate_capdirty), as there's certainly no reason to use it in the user PC case.